kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/ResponsiveConsumer.java

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright 2023 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.clients;
+
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.regex.Pattern;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerGroupMetadata;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ResponsiveConsumer<K, V> implements Consumer<K, V> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResponsiveConsumer.class);
+
+  private final Consumer<K, V> wrapped;
+  private final List<Listener> listeners;
+
+  private class RebalanceListener implements ConsumerRebalanceListener {
```

Comment:
I can make it static - though I don't really understand the reference leaking concern here. I really don't think it should be package-private as it's strictly an implementation detail of this class.

Reply:
the reference leaking of non-static inner classes means that if another class holds onto this it will also have a reference to the parent class, which can make garbage collection leaks difficult to debug - I believe this is passed to the other classes as part of the rebalance listener API, right?

Reply:
Yeah, it's passed and held internally within the consumer in the objects that maintain metadata about the subscription. There's definitely a cycle in the reference chain. Something like:

stream thread -> responsive consumer -> consumer -> subscription metadata -> responsive rebalance listener -> responsive consumer.

But my understanding was that's fine - they will all get cleaned up once the stream thread drops the reference to the consumer (or the stream thread goes away).

So I'm still not understanding the concern. Not trying to be argumentative (I've already changed it to static) - I'm just genuinely curious how a non-static class could cause an issue with leak - it's not something I've run into or considered before.